src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp

```diff
@@ -16,6 +17,99 @@
+ * scores using inverse normal transformation and a fractional offset. Based on
+ * paper https://arxiv.org/abs/1903.08008
+ * @param draws stores chains in columns
+ * @return normal scores for average ranks of draws
+ *
+ */
+
+Eigen::MatrixXd rank_transform(const Eigen::MatrixXd& draws) {
+  int rows = draws.rows();
+  int cols = draws.cols();
+  int size = rows * cols;
+  Eigen::MatrixXd rankMatrix = Eigen::MatrixXd::Zero(rows, cols);
+
+  std::vector<std::pair<double, int>> valueWithIndex(size);
+
+  for (int col = 0; col < cols; ++col) {
+    for (int row = 0; row < rows; ++row) {
+      int index
+          = col * rows + row;  // Calculating linear index in column-major order
+      valueWithIndex[index] = {draws(row, col), index};
+    }
+  }
+
+  std::sort(valueWithIndex.begin(), valueWithIndex.end());
+
+  // Assigning average ranks
+  for (int i = 0; i < size; ++i) {
+    // Handle ties by averaging ranks
+    int j = i;
+    double sumRanks = 0;
+    int count = 0;
+
+    while (j < size && valueWithIndex[j].first == valueWithIndex[i].first) {
+      sumRanks += j + 1;  // Rank starts from 1
+      ++j;
+      ++count;
+    }
+
+    double avgRank = sumRanks / count;
+    boost::math::normal_distribution<double>
+        dist;  // Standard normal distribution
+    for (int k = i; k < j; ++k) {
+      int index = valueWithIndex[k].second;
+      int row = index % rows;  // Adjusting row index for column-major order
+      int col = index / rows;  // Adjusting column index for column-major order
+      double p = (avgRank - 3.0 / 8.0) / (size - 2.0 * 3.0 / 8.0 + 1.0);
+      rankMatrix(row, col) = boost::math::quantile(dist, p);
+    }
+    i = j - 1;  // Skip over tied elements
+  }
+  return rankMatrix;
+}
+
+/**
+ * Computes square root of marginal posterior variance of the estimand by
+ * weigted average of within-chain variance W and between-chain variance B.
+ *
+ * @param draws stores chains in columns
+ * @return square root of ((N-1)/N)W + B/N
+ *
+ */
+inline double rhat(const Eigen::MatrixXd& draws) {
+  int num_chains = draws.cols();
+  int num_draws = draws.rows();
+  Eigen::VectorXd chain_mean(num_chains);
+  boost::accumulators::accumulator_set<
+      double, boost::accumulators::stats<boost::accumulators::tag::variance>>
+      acc_chain_mean;
+  Eigen::VectorXd chain_var(num_chains);
+  double unbiased_var_scale = num_draws / (num_draws - 1.0);
+  for (int chain = 0; chain < num_chains; ++chain) {
+    boost::accumulators::accumulator_set<
+        double, boost::accumulators::stats<boost::accumulators::tag::mean,
+                                           boost::accumulators::tag::variance>>
+        acc_draw;
+    for (int n = 0; n < num_draws; ++n) {
+      acc_draw(draws(n, chain));
+    }
+    chain_mean(chain) = boost::accumulators::mean(acc_draw);
+    acc_chain_mean(chain_mean(chain));
+    chain_var(chain)
+        = boost::accumulators::variance(acc_draw) * unbiased_var_scale;
+  }
```

Comment:
It doesn't look like you need an online way to update and view the current mean and variance. You can calculate the mean for each chain via `draws.colwise().mean()` and `draws.array().mean()` for the overall mean. Then a loop for the chain variance calculation.

Now that I'm reading this again should that be named `acc_chain_variance because of line 105? I've never used boost accumulators before so not sure if I'm following all the way here